Draft & discussion: SquareLine generator (#56)
espressif/esp-bsp: SquareLine/README.md

# SquareLine board generator

> This is simple python generator for SquareLine board packages. This generator copy all files from `common` folder to destination ZIP file and generates other files by `manifest.json` file saved in each board directory in `boards` folder. 

Reviewer: ```suggestion
This is a simple python generator script for SquareLine board packages. This script copies all files from `common` folder to the destination ZIP file. It also generates other files according to the `manifest.json` file from each board directory in `boards` folder. For information on the structure of SquareLine board packages, please refer to (add link here).
```

## Mandatory files in each board folder:
* `manifest.json` - Description of the board
* `image.png` - Image of the board 380px x 300px
* `sdkconfig.defaults` - defaults
* `main/idf_component.yml` - Component Registry description

### `manifest.json`

Mandatory JSON keys in manifest:
* `name` - Board name (string)

> * `version` - Package version (string)

Reviewer: Should we add a note, when should this version be modified?

* `mcu` - ESP mcu (string)
* `screen_width` - Screen width (string)
* `screen_height` - Screen height (string)
* `screen_color_swap` - Is color bytes swapped (boolean)
* `short_description` - Short board description (string)
* `long_description` - Long board description (string)
* `placeholders` - List of placeholders and values of them (JSON object)

### Placeholders

> During the copy all files (except `image.png`), there can be replaced some placeholder by the value. Placeholder must be in compound brackets (e.g. `{PLACEHOLDER}`). The values for placeholders are saved in `manifest.json` file without compound brackets like this:

Reviewer: ```suggestion
When copying all the files (except for `image.png`), the script will replace placeholders found in these files. Placeholder must be in compound brackets (e.g. `{PLACEHOLDER}`). The placeholders and their values should be defined in `manifest.json` as follows:
```

```
    "placeholders":
    {
        "PLACEHOLDER_1": "value 1",
        "PLACEHOLDER_2": "value 2"
    }
```

> **_NOTE 1:_** Placeholders are optional.

Reviewer: I'm not sure what "optional" means here. The note below says that if the placeholder is not defined in Json, it will not be replaced in the source file. So it looks like it's not optional?

Author: It means, that all JSON keys must be there except placeholders. When there is missing some other key, the script fails. When there is missing some placeholder, it continues, but not replace the placeholder.

> **_NOTE 2:_** When some placeholder is found in code and not found in JSON, it will leave there in a same form.

Reviewer: ```suggestion
**_NOTE 2:_** When some placeholder is found in code and not found in JSON, it will not be substituted.
```

By the way, is this behavior intentional, is it to catch typos in placeholder names?

Author: Yes, I want it. Because there can be used lot of other placeholders from other systems like IDF (e.g. {IDF_PATH} in CMakeLists) and these shouldn't be replaced.
